### ISSUE TITLE
Iteration-based training

### ISSUE THREAD

**Issue by janvainer** [catalyst-team/catalyst#913]
## 🚀 Feature Request
Make iteration-based training available - instead of specifying epoch count, the user should be ableto specify gradient step cound and spefify how many steps form an epoch. The dataloader would be cycled indefinitely.

### Motivation
Lets say I have a very large dataset. One epoch of training (seeing all the data) can easily take several hours.
I want to be able to do the following actions after a certain number of seen batches irrespective of wether the epoch is over or not:

1. Run validation on validation set
2. Show example outputs
3. Make scheduler step - this is very important, because If I run only say 4 epochs on my large dataset, I get only 4 scheduler steps, but the optimizer has made very many steps in the mean time
4. Save a checkpoint

### Proposal
I would like an interface to be something like this:

```
    runner.train(
        model, 
        train_loader,
        epochs=args.epochs,
        scheduler_step_every=1,   # each epoch, make a step
        steps_per_epoch=args.steps_per_epoch,  # how many gradient steps form an epoch
        valid_loader=valid_loader,
    )
```

 It would be nice to have this option in the default api.
The main problem is how to make the dataloader go infinitely. One can wrap any dataloader in something like Infinite Stream:
```
class InfiniteEpochStream:
    def __init__(self, data_loader, steps_per_epoch):
        self.data_loader = data_loader
        self.steps_per_iter = steps_per_epoch
        self.iterator = iter(self.data_loader)

        self.__steps = 0
        self.__iterator_steps = 0

    def __iter__(self):
        self.__steps = 0
        return self

    def __len__(self):
        return self.steps_per_iter

    def __next__(self):
        if self.__steps < self.steps_per_iter:
            if self.__iterator_steps >= len(self.data_loader):
                self.__refresh_iterator()

            self.__steps += 1
            self.__iterator_steps += 1
            return next(self.iterator)
        else:
            raise StopIteration

    def __refresh_iterator(self):
        self.iterator = iter(self.data_loader)
        self.__iterator_steps = 0
```
A dataloader wrapped in this class should cycle indefinitely over random batches.
Now we can do the following:

```
num_samples, num_features, num_classes = int(1e4), int(1e1), 4
X = torch.rand(num_samples, num_features)
y = (torch.rand(num_samples, ) * num_classes).to(torch.int64)

# pytorch loaders
dataset = TensorDataset(X, y)
loader = DataLoader(dataset, batch_size=32, num_workers=1)
loaders = {
    "train": InfiniteEpochStream(loader, steps_per_epoch=1000),  # after 1000 steps, the loader wil return StopIteration
    "valid": loader}

# model, criterion, optimizer, scheduler
model = torch.nn.Linear(num_features, num_classes)
criterion = torch.nn.CrossEntropyLoss()
optimizer = torch.optim.Adam(model.parameters())
scheduler = torch.optim.lr_scheduler.MultiStepLR(optimizer, [2])

# model training
runner = dl.SupervisedRunner()
runner.train(
    model=model,
    criterion=criterion,
    optimizer=optimizer,
    scheduler=scheduler,
    loaders=loaders,
    logdir="./logdir",
    num_epochs=3,
    callbacks=[dl.AccuracyCallback(num_classes=num_classes)]
)
```

I am not sure how this would play with the distributed training as the [distributed training examples ](https://catalyst-team.github.io/catalyst/info/distributed.html) do not specify Dataloaders but use datasets with params for the default dataloader.
